sbncode/BeamSpillInfoRetriever/BNBRetriever/BNBRetriever_module.cc

```diff
@@ -0,0 +1,320 @@
+  void beginSubRun(art::SubRun& sr);
+  void endSubRun(art::SubRun& sr);
+
+private:
+  // input labels
+  std::vector< sbn::BNBSpillInfo > fOutbeamInfos;
+  int fTimePad;
+  std::string fURL;
+  MWRData mwrdata;
+  std::string raw_data_label_;
+  std::string fDeviceUsedForTiming;
+  int TotalBeamSpills;  
+  //
+  art::ServiceHandle<ifbeam_ns::IFBeam> ifbeam_handle;
+  std::unique_ptr<ifbeam_ns::BeamFolder> bfp;
+  std::unique_ptr<ifbeam_ns::BeamFolder> bfp_mwr;
+
+
+
+};
+
+
+sbn::BNBRetriever::BNBRetriever(fhicl::ParameterSet const& p)
+  : EDProducer{p},
+  fTimePad(p.get<double>("TimePadding",0.0333)), //seconds 
+  raw_data_label_(p.get<std::string>("raw_data_label")),
+  fDeviceUsedForTiming(p.get<std::string>("DeviceUsedForTiming")),
+  bfp(     ifbeam_handle->getBeamFolder(p.get< std::string >("Bundle"), p.get< std::string >("URL"), p.get< double >("TimeWindow"))),
+  bfp_mwr( ifbeam_handle->getBeamFolder(p.get< std::string >("MultiWireBundle"), p.get< std::string >("URL"), p.get< double >("TimeWindow")))
+{
+ 
+  // how close in time does the spill time have to be from the DAQ time (in seconds).
+  // If these are too large then it fails to capture the device 
+  // If these are too small then the time jitter in devices means we miss good data 
+  bfp->set_epsilon(0.02); //20 ms, this was tuned by hand and compared to IFBeamDB times  
+  
+  bfp_mwr->set_epsilon(100); // TO BE TUNED!
+  produces< std::vector< sbn::BNBSpillInfo >, art::InSubRun >();
+  TotalBeamSpills = 0;
+}
+
+void sbn::BNBRetriever::produce(art::Event& e)
+{
+  
+  //Here we read in the artdaq Fragments and extract three pieces of information:
+  // 1. The time of the current event, t_current_event
+  // 2. the time of the previously triggered event, t_previous_event (NOTE: Events are non-sequential!)
+  // 3. the number of beam spills since the previously triggered event, number_of_gates_since_previous_event
+  
+  int gate_type = 0;
+  art::Handle< std::vector<artdaq::Fragment> > raw_data_ptr;
+  e.getByLabel(raw_data_label_, "ICARUSTriggerUDP", raw_data_ptr);
+  auto const & raw_data = (*raw_data_ptr);
+
+  double t_current_event  = 0;
+  double t_previous_event = 0;
+  double number_of_gates_since_previous_event = 0;
+  
+  for(auto raw_datum : raw_data){
+   
+    uint64_t artdaq_ts = raw_datum.timestamp();
+    icarus::ICARUSTriggerUDPFragment frag(raw_datum);
+    std::string data = frag.GetDataString();
+    char *buffer = const_cast<char*>(data.c_str());
+    icarus::ICARUSTriggerInfo datastream_info = icarus::parse_ICARUSTriggerString(buffer);
+    gate_type = datastream_info.gate_type;
+    number_of_gates_since_previous_event = frag.getDeltaGatesBNB();
+  
+    t_current_event = static_cast<double>(artdaq_ts)/(1000000000); //check this offset...
+    if(gate_type == 1)
+      t_previous_event = (static_cast<double>(frag.getLastTimestampBNB()))/(1e9);
+    else
+      t_previous_event = (static_cast<double>(frag.getLastTimestampOther()))/(1000000000);
```

Comment:
this is basically just a placeholder, but I think it is because our "proof of principle" file had the wrong Numi gate-type assigned. This will be deleted long term since we only care about BNB for this module

Reply:
Ok... if things move fast enough with the other side, before this code is merged you will be dealing already with the new data product which will hide this issue (i.e. _its_ producer will need to take care of it).